### ISSUE TITLE
Bug: [no-unnecessary-parameter-property-assignment] false positive after reassigning a constructor parameter

### ISSUE THREAD

**Issue by zaewc** [typescript-eslint/typescript-eslint#12834]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.
- [x] If I used AI assistance to prepare this issue, I have reviewed and followed the [AI Contribution Policy](https://typescript-eslint.io/contributing/ai-policy/).

### Playground Link

https://typescript-eslint.io/play/#fileType=.ts&code=MYGwhgzhAECqEFMBO0DeAoa1gHsB2EALkgK7CE5IAUADiQEYgCWw0eYAtggFzRFJM8AcwCUaTFjacE0ALxSuAOmJMOVEQG4JWQgAsmERey5yFCLVgC%2B6a%2BlwEcIBIpA4hVPAgDucRNQDk0ACCzMAy-iJG0ppAA&eslintrc=N4IgTgrgNgpgziAXKAAgFwJ4Ad4GMwCWWaAtPFAQHZoD0lA9iRJZTLvHAIZgYlbecAtjDQwwfMPRxhMJTnDgEA5pWHUkIMZLAgAvrqA&tsconfig=N4Igxg9gtgDglgGwKYCcDyMAucIDsDOIAXKPpinGJseQK5IC%2BDQA

### Repro Code

```TypeScript
class User {
  constructor(public name: string) {
    name = name.trim();
    this.name = name;
  }
}

console.log(new User(' Alice ').name);
```

### ESLint Config

```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  rules: {
    '@typescript-eslint/no-unnecessary-parameter-property-assignment': 'error',
  },
};
```

### tsconfig

```jsonc
na
```

### Expected Result

No report 

### Actual Result

The rule reports the assignment as unnecessary. Removing it changes the output from `'Alice'` to `' Alice '`

**Comment by diptomistry** [comment]
taking a look
